libqtile/widget/battery.py

```diff
@@ -267,7 +277,64 @@ def _get_param(self, name) -> tuple[str, str]:
 
         raise RuntimeError("Unable to read status for {}".format(name))
 
+    def connected_to_thunderbolt(self):
+        try:
+            sysfs = "/sys/bus/thunderbolt/devices"
+            entries = os.listdir(sysfs)
+            for e in entries:
+                try:
+                    name = Path(sysfs, e, "device_name").read_text()
+                except FileNotFoundError:
+                    continue
+                else:
+                    logger.debug("found dock %s", name)
+                    return True
+        except OSError:
+            logger.debug("failed to detect thunderbot %s", exc_info=True)
+        return False
+
+    def set_battery_charge_thresholds(self):
+        if not self.charge_threshold_supported:
+            return
+
+        start = 0
+        end = self.default_max_charge
+
+        # if we are thunderbolt docked, set the thresholds to 40/50, per
+        # https://support.lenovo.com/us/en/solutions/ht078208-how-can-i-increase-battery-life-thinkpad-and-lenovo-vbke-series-notebooks
+        if self.connected_to_thunderbolt():
+            start = 40
+            end = 50
+
+        # unless we have been explicitly asked to charge, then we should JFDI
+        if self.force_charge:
+            start = 0
+            end = 100
+
+        battery_dir = "/sys/class/power_supply"
+
+        path = os.path.join(battery_dir, self.battery, "charge_control_start_threshold")
+        try:
+            with open(path, "w+") as f:
+                f.write(str(start))
+        except FileNotFoundError:
+            self.charge_threshold_supported = False
+        except OSError:
+            self.charge_threshold_supported = False
+            logger.exception("Failed to write %s (did you set up udev rules?)", path)
+
+        path = os.path.join(battery_dir, self.battery, "charge_control_end_threshold")
+        try:
+            with open(path, "w+") as f:
+                f.write(str(end))
+        except FileNotFoundError:
+            self.charge_threshold_supported = False
+        except OSError:
+            self.charge_threshold_supported = False
+            logger.exception("Failed to write %s (did you set up udev rules?)", path)
+
     def update_status(self) -> BatteryStatus:
+        self.set_battery_charge_thresholds()
```

Comment:
Should there be an option to toggle whether we want smart charging enabled or not, rather than just with the `qtile cmd-obj` commands?

Reply:
Perhaps just letting people set the default value of force_charge?